### PR TITLE
PIM-3591: introduce IdentifiableObjectRepositoryInterface

### PR DESCRIPTION
**Pull request by jjanvier**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | N
| BC breaks?           | only 1 made on purpose :)
| CI currently passes? | Y
| Tests pass?          | Y
| Scenarios pass?      | ORM Y / ODM N
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | Y
| Fixed tickets        |
| DB schema updated?   |
| Migration script?    |
| Doc PR               |

This PR introduces : 
 - `Akeneo/Bundle/StorageUtilsBundle/Doctrine/ORM/Repository/IdentifiableEntityRepository`
 - `Akeneo/Bundle/StorageUtilsBundle/Repository/AbstractIdentifiableEntityRepository`

*... while assuring backward compatibility*

Please, note that for backward compatibility reasons : 
 - we can not directly remove `ReferableEntityRepositoryInterface` and `ReferableEntityRepository`, otherwise no existing PIM customization will continue to work after the upgrade
 - the repositories that implemented `ReferableEntityRepositoryInterface` now implements `ReferableEntityRepositoryInterface` ***and*** `IdentifiableObjectRepositoryInterface`. The implementation of `ReferableEntityRepositoryInterface` will be set to deprecated in another PR (see PR https://github.com/akeneo/pim-community-dev/pull/2022 where all repositories will have an interface). The methods implemented for `ReferableEntityRepositoryInterface` are just simple shortcuts to methods implemented for `IdentifiableObjectRepositoryInterface`.
 - the repositories that extends `ReferableEntityRepository` ***do*** continue to extend `ReferableEntityRepository`. `ReferableEntityRepository` is simply set as deprecated for the moment. We'll encourage users to rely on repository interfaces instead of concrete classes (see PR https://github.com/akeneo/pim-community-dev/pull/2022 where all repositories will have an interface). 
 - we have to check here and there if a repository is an `IdentifiableObjectRepositoryInterface` or a `ReferableEntityRepositoryInterface`, which introduces bad code. This bad code has been isolated in private functions.

### CODE REVIEW

File: src/Akeneo/Bundle/StorageUtilsBundle/Repository/IdentifiableObjectRepositoryInterface.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace Akeneo\Bundle\StorageUtilsBundle\Repository;
+
+/**
+ * Interface for repositories of unique code objects
+ *
+ * @author    Antoine Guigan <antoine@akeneo.com>
+ * @copyright 2013 Akeneo SAS (http://www.akeneo.com)
```
Comment (nidup) [inline src/Akeneo/Bundle/StorageUtilsBundle/Repository/IdentifiableObjectRepositoryInterface.php]: author and copyright

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ArrayToObject/AbstractProcessor.php [hunk 2]
```diff
@@ -78,14 +78,15 @@ public function setStepExecution(StepExecution $stepExecution)
...
-     * @param string                             $class      the class to instanciate in case the
-     *                                                       object has not been found
+     * @param IdentifiableObjectRepositoryInterface $repository the repository to search inside
+     * @param array                                 $data       the data that is currently processed
+     * @param string                                $class      the class to instanciate in case the
+     *                                                          object has not been found
+
      *
-     * @return object
+*@return object
```
Comment (nidup) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ArrayToObject/AbstractProcessor.php]: alignement

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepository.php [hunk 3]
```diff
@@ -251,24 +253,7 @@ protected function getFullProductQB()
     /**
      * {@inheritdoc}
      */
-    public function findByReference($code)
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepository.php]: Looks weird to drop this to use parent findOneByIdentifier, imho it doesnt work

File: src/Pim/Bundle/TransformBundle/Builder/FieldNameBuilder.php [hunk 4]
```diff
@@ -236,8 +239,11 @@ protected function checkFieldNameLocaleByChannel(AttributeInterface $attribute,
             isset($attributeInfos['scope_code']) &&
             isset($attributeInfos['locale_code'])
         ) {
-            $channel = $this->getRepository($this->channelClass)->findByReference($attributeInfos['scope_code']);
-            $locale = $this->getRepository($this->localeClass)->findByReference($attributeInfos['locale_code']);
+            $channelRepository = $this->getRepository($this->channelClass);
+            $localeRepository = $this->getRepository($this->localeClass);
+
+            $channel = $this->findOneByIdentifier($channelRepository, $attributeInfos['scope_code']);
```
Comment (nidup) [inline src/Pim/Bundle/TransformBundle/Builder/FieldNameBuilder.php]: extra parameter never used here

File: src/Pim/Bundle/TransformBundle/Builder/FieldNameBuilder.php [hunk 5]
```diff
@@ -236,8 +239,11 @@ protected function checkFieldNameLocaleByChannel(AttributeInterface $attribute,
...
+            $locale = $this->findOneByIdentifier($localeRepository, $attributeInfos['locale_code']);
```
Comment (nidup) [inline src/Pim/Bundle/TransformBundle/Builder/FieldNameBuilder.php]: extra parameter never used here $attributeInfos['locale_code']

File: src/Pim/Bundle/UserBundle/Entity/Repository/GroupRepository.php [hunk 6]
```diff
@@ -52,14 +54,35 @@ public function getAllButDefaultQB()
...
 
     /**
      * {@inheritdoc}
      */
-    public function getReferenceProperties()
+    public function getIdentifierProperties()
     {
         return array('name');
     }
+
```
Comment (nidup) [inline src/Pim/Bundle/UserBundle/Entity/Repository/GroupRepository.php]: useless blank line, please use cs fixer

Comment (jjanvier) [inline src/Pim/Bundle/UserBundle/Entity/Repository/GroupRepository.php]: doen

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepository.php (hunk 3 above)
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepository.php]: @nidup two methods already existed `findByReference` and `findOneByIdentifier`. There implementation were different but the goal was the same. I kept `findOneByIdentifier` as its implementation is more generic. I deleted the "old" `findByReference`. Now the new `findByReference` is a shortcut to `findOneByIdentifier` that was already present here

File: src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/ORM/Cursor.php [hunk 7]
```diff
@@ -143,13 +143,11 @@ protected function getRepository()
...
-            if (!($this->repository instanceof ModelRepositoryInterface)) {
-                throw new LogicException(
-                    sprintf(
-                        '%s repository must implement ModelRepositoryInterface',
-                        $entityClass
-                    )
-                );
+            if (!$this->repository instanceof CursorableRepositoryInterface) {
+                $message = <<<MESSAGE
+'"%s" repository must implement "Akeneo\Bundle\StorageUtilsBundle\Doctrine\ORM\Repository\CursorableRepositoryInterface"'
```
Comment (nidup) [inline src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/ORM/Cursor.php]: concatenate with a `.`

File: src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/ORM/Repository/AbstractIdentifiableEntityRepository.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+use Doctrine\ORM\EntityRepository;
+
+/**
+ * Base repository for entities with a code unique index
+ *
+ * @author    Julien Janvier <jjanvier@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+abstract class AbstractIdentifiableEntityRepository extends EntityRepository implements
```
Comment (nidup) [inline src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/ORM/Repository/AbstractIdentifiableEntityRepository.php]: we could drop that, as discussed together more clear if we implement the 2 methods

File: src/Pim/Bundle/TransformBundle/Cache/DoctrineCache.php [hunk 9]
```diff
@@ -108,13 +111,39 @@ protected function findObject($class, $code)
...
-                throw new \Exception(
-                    sprintf('Repository "%s" of class "%s" is not referable', get_class($repository), $class)
-                );
+            if (null !== $object = $this->findOneByIdentifier($repository, $code)) {
+                return $object;
             }
 
-            return $repository->findByReference($code);
+            $message = <<<MESSAGE
+Repository "%s" of class "%s" does not implement "Akeneo\Bundle\StorageUtilsBundle\Repository\IdentifiableObjectRepositoryInterface"
```
Comment (nidup) [inline src/Pim/Bundle/TransformBundle/Cache/DoctrineCache.php]: concatenate with `.`

File: src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/ORM/Cursor.php (hunk 7 above)
Comment (jjanvier) [inline src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/ORM/Cursor.php]: done

File: src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/ORM/Repository/AbstractIdentifiableEntityRepository.php (hunk 8 above)
Comment (jjanvier) [inline src/Akeneo/Bundle/StorageUtilsBundle/Doctrine/ORM/Repository/AbstractIdentifiableEntityRepository.php]: done

File: src/Pim/Bundle/TransformBundle/Cache/DoctrineCache.php (hunk 9 above)
Comment (jjanvier) [inline src/Pim/Bundle/TransformBundle/Cache/DoctrineCache.php]: done

**Pull request merged**